### PR TITLE
Update vector to a similar API as map. Add some unittests

### PR DESCRIPTION
**Pull request by wilsonk**
Added some comments for previous unittests, just for easier info retrieval on all of them, as there are a pretty good amount of total tests now. 

Added the toHash functionality requested in #25

Also, I changed the name of the 'has()' function to 'contains()' to match the map implementation...I can change the name of both back to 'has', if needed, but the comment literally said 'return true if vector contains value'...so this change seemed to make sense.

I tried using the iterators from iterator.d in vector.d, but I ran in to troubles with return values (might need to use pointers), and different constructs for map and vector (pair vs single item)...so I am not totally sure how to handle this generically. Any thoughts?

I might look at the iterators again when I get around to updating the other containers to a similar API. I think I removed all the asserts in the code and only the tests have asserts in this PR.

Thanks.

### CODE REVIEW

**Comment by nxuv** [comment]
Change "contains" back to "has" please, just a personal preference. But do add contains as an alias to has

File: src/clib/vector.d [hunk 1]
```diff
@@ -38,32 +38,144 @@ struct vector(T, A: IAllocator!T = allocator!T) if (!is(T == bool)) {
...
+    ConstReverseVectorIterator!(T) rbegin() const @nogc nothrow {
+        return ConstReverseVectorIterator!(T)(&this, false);
+    }
+
+    /// Returns const backward iterator to end (for reverse iteration)
+    ConstReverseVectorIterator!(T) rend() const @nogc nothrow {
+        return ConstReverseVectorIterator!(T)(&this, true);
+    }
+
+    private struct VectorIterator(T) {
```
Comment (nxuv) [inline src/clib/vector.d:85]: I'd say prefer using iterators from clib.iterator and avoid declaring private types that would be returned from functions

File: src/clib/vector.d [hunk 2]
```diff
@@ -180,7 +332,7 @@ struct vector(T, A: IAllocator!T = allocator!T) if (!is(T == bool)) {
     }
 
     /// Returns true if vector contains value
-    bool has(T val) @nogc nothrow {
+    bool contains(T val) @nogc nothrow {
```
Comment (nxuv) [inline src/clib/vector.d:335]: as I said in first comment, revert back and create alias

File: src/clib/vector.d [hunk 3]
```diff
@@ -547,83 +731,129 @@ struct vector(T, A: IAllocator!T = allocator!T) if (!is(T == bool)) {
...
 
 // Unittests
-// TODO: more edge cases
 @nogc nothrow {
 
+    // Test vector initialization from array and variadic constructor
     unittest {
-        int[3] testArr = [2, 5, 6];
+        int[3] testArr;
+        testArr[0] = 2; testArr[1] = 5; testArr[2] = 6;
```
Comment (nxuv) [inline src/clib/vector.d:739]: why splitting?

File: src/clib/vector.d [hunk 4]
```diff
@@ -547,83 +731,129 @@ struct vector(T, A: IAllocator!T = allocator!T) if (!is(T == bool)) {
...
         assert(v.array == [1, 4, 2]);
     }
 
+    // Test vector push operations and array concatenation
     unittest {
-        int[3] a = [1, 2, 3];
+        int[3] a;
+        a[0] = 1; a[1] = 2; a[2] = 3;
         vector!int v = a;
-        v ~= 2;
```
Comment (nxuv) [inline src/clib/vector.d:565]: same, why not use existing ops?

**Comment by nxuv** [comment]
As for iterators they are generic, that's what IS_VALID_ITERABLE is for, so... idk what's the issue

Looked at map and there's same privately defined iterator, not good
